pkg/deploy/registry/deploylog/rest.go

```diff
@@ -0,0 +1,237 @@
+	if err != nil {
+		return nil, err
+	}
+	if len(deployments.Items) == 0 {
+		return nil, errors.NewInternalError(fmt.Errorf("cannot find deployments for deploymentConfig %q", config.Name))
+	}
+	sort.Sort(deployutil.DeploymentsByLatestVersionDesc(deployments.Items))
+	// Use latest deployment
+	target := &deployments.Items[0]
+	if desiredVersion != deployutil.DeploymentVersionFor(target) {
+		glog.V(3).Infof("Cannot find latest deployment for %q, using %q", config.Name, target.Name)
+	}
+
+	status := deployutil.DeploymentStatusFor(target)
+	switch status {
+	case deployapi.DeploymentStatusNew, deployapi.DeploymentStatusPending:
+		if deployLogOpts.NoWait {
+			glog.V(4).Infof("Deployment %s/%s is in %s state. No logs to retrieve yet.", target.Namespace, target.Name, status)
+			return &genericrest.LocationStreamer{}, nil
+		}
+		glog.V(4).Infof("Deployment %s is in %s state, waiting for it to start...", deployutil.LabelForDeployment(target), status)
+		lw := createDeploymentLW(ctx, r.Deploy, target)
+		if err := WaitForResource(&kapi.ReplicationController{}, lw, handleDeployment); err != nil {
+			return nil, err
+		}
+	case deployapi.DeploymentStatusComplete:
+		// Deployer pod has been deleted, no logs to retrieve
+		glog.V(4).Infof("Deployment %s/%s was successful so the deployer pod is deleted. No logs to retrieve.", target.Namespace, target.Name)
+		return &genericrest.LocationStreamer{}, nil
+	}
+
+	deployPodName := deployutil.DeployerPodNameForDeployment(target.Name)
+	logOpts := &kapi.PodLogOptions{
+		Follow: deployLogOpts.Follow,
+	}
+	location, transport, err := pod.LogLocation(r.PodGetter, r.ConnectionInfo, ctx, deployPodName, logOpts)
+	if err != nil {
+		return nil, errors.NewBadRequest(err.Error())
+	}
+	return &genericrest.LocationStreamer{
+		Location:    location,
+		Transport:   transport,
+		ContentType: "text/plain",
+		Flush:       deployLogOpts.Follow,
+	}, nil
+}
+
+// Resource specific stuff
+
+// createDeploymentLW creates a ListWatch scoped on deployments
+func createDeploymentLW(ctx kapi.Context, c *Client, deployment *kapi.ReplicationController) *cache.ListWatch {
+	return &cache.ListWatch{
+		ListFunc: func() (runtime.Object, error) {
+			return c.ListDeployments(ctx, labels.Everything())
+		},
+		WatchFunc: func(rv string) (watch.Interface, error) {
+			return c.WatchDeployments(ctx, labels.Everything(), fields.Set{"metadata.name": deployment.Name}.AsSelector(), rv)
+		},
+	}
+}
+
+// handleDeployment is a ReactFn scoped on deployments
+func handleDeployment(obj interface{}) (bool, error) {
+	deployment, ok := obj.(*kapi.ReplicationController)
+	if !ok {
+		return false, fmt.Errorf("event object is not a deployment: %#v", obj)
+	}
+
+	switch deployutil.DeploymentStatusFor(deployment) {
+	case deployapi.DeploymentStatusRunning, deployapi.DeploymentStatusComplete, deployapi.DeploymentStatusFailed:
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// Generic stuff
```

Comment:
My takeaway from the discussion with @smarterclayton was to just use a `watch.Interface`.